gio-subclass/src/application.rs

```diff
@@ -0,0 +1,599 @@
+        application.parent_quit_mainloop();
+    }
+
+    fn shutdown(&self, application: &T) {
+        application.parent_shutdown();
+    }
+
+
+    // fn handle_local_options(&self, application: &T, options: &glib::VariantDict) {
+    //     application.handle_local_options(options);
+    // }
+}
+
+pub trait ApplicationImplExt<T> {}
+
+impl<S: ApplicationImpl<T>, T: ObjectType + glib::IsA<gio::Application>> ApplicationImplExt<T>
+    for S
+{
+}
+
+any_impl!(ApplicationBase, ApplicationImpl);
+
+pub unsafe trait ApplicationBase: IsA<gio::Application> + ObjectType {
+    fn parent_startup(&self) {
+        unsafe {
+            let klass = self.get_class();
+            let parent_klass = (*klass).get_parent_class() as *const gio_ffi::GApplicationClass;
+            (*parent_klass)
+                .startup
+                .map(|f| f(self.to_glib_none().0))
+                .unwrap_or(())
+        }
+    }
+
+    fn parent_activate(&self) {
+        unsafe {
+            let klass = self.get_class();
+            let parent_klass = (*klass).get_parent_class() as *const gio_ffi::GApplicationClass;
+            (*parent_klass)
+                .activate
+                .map(|f| f(self.to_glib_none().0))
+                .unwrap_or(())
+        }
+    }
+
+    fn parent_open(&self, files: &Vec<gio::File>, hint: &String) {
+        unsafe {
+            let klass = self.get_class();
+            let parent_klass = (*klass).get_parent_class() as *const gio_ffi::GApplicationClass;
+            (*parent_klass)
+                .open
+                .map(|f| {
+                    let n_files = files.len() as i32;
+                    f(
+                        self.to_glib_none().0,
+                        files.to_glib_none().0,
+                        n_files,
+                        hint.to_glib_none().0,
+                    )
+                })
+                .unwrap_or(())
+        }
+    }
+
+    fn parent_command_line(&self, cmd_line: &gio::ApplicationCommandLine) -> i32 {
+        unsafe {
+            let klass = self.get_class();
+            let parent_klass = (*klass).get_parent_class() as *const gio_ffi::GApplicationClass;
+            (*parent_klass)
+                .command_line
+                .map(|f| f(self.to_glib_none().0, cmd_line.to_glib_none().0))
+                .unwrap_or(0)
+        }
+    }
+
+    fn parent_local_command_line(&self, arguments: &mut Vec<String>) -> Option<i32> {
+        unsafe {
+            let klass = self.get_class();
+            let parent_klass = (*klass).get_parent_class() as *const gio_ffi::GApplicationClass;
+            let mut exit_status = 0;
+            let success = (*parent_klass)
+                .local_command_line
+                .map(|f| {
+                    let mut args = arguments.to_glib_full();
+                    f(
+                        self.to_glib_none().0,
+                        &mut args,
+                        &mut exit_status,
+                    )
```

Comment:
You need to check afterwards if args were changed and update the arguments vector. The docs say:

```
   * This virtual function is always invoked in the local instance. It
   * gets passed a pointer to a %NULL-terminated copy of @argv and is
   * expected to remove arguments that it handled (shifting up remaining
   * arguments).
```

So this means that the array pointer itself does not change, but the items in the array might be changed (items might be removed only).

You also need to free the args pointer afterwards (pass the pointer to glib_ffi::g_strfreev)